apps/emqx_durable_storage/src/emqx_ds_beamformer_catchup.erl

```diff
@@ -0,0 +1,429 @@
+                    ),
+                    handle_recoverable(S, [Req]);
+                Err = {error, unrecoverable, Reason} ->
+                    ?tp(
+                        error,
+                        emqx_ds_beamformer_move_to_rt_fail,
+                        #{shard => DBShard, recoverable => false, req => Req, error => Reason}
+                    ),
+                    emqx_ds_beamformer:send_out_final_beam(DBShard, SubTab, Err, [Req]),
+                    queue_drop(Queue, Req)
+            end
+        end,
+        Reqs
+    ).
+
+-spec fulfill_batch(
+    s(),
+    emqx_ds:stream(),
+    emqx_ds:topic_filter(),
+    emqx_ds:message_key(),
+    emqx_ds:message_key(),
+    [{emqx_ds:message_key(), emqx_types:message()}]
+) -> ok.
+fulfill_batch(
+    S = #s{shard_id = ShardId, sub_tab = SubTab, metrics_id = Metrics, module = CBM, queue = Queue},
+    Stream,
+    TopicFilter,
+    StartKey,
+    EndKey,
+    Batch
+) ->
+    %% Find iterators that match the start message of the batch (to
+    %% handle iterators freshly created by `emqx_ds:make_iterator'):
+    Candidates = queue_lookup(S, Stream, TopicFilter, StartKey),
+    %% Search for iterators where `last_seen_key' is equal to key of
+    %% any message in the batch:
+    BeamMaker =
+        process_batch(
+            S,
+            Stream,
+            TopicFilter,
+            Batch,
+            Candidates,
+            emqx_ds_beamformer:beams_init(
+                CBM,
+                ShardId,
+                SubTab,
+                true,
+                fun(Req) -> queue_drop(Queue, Req) end,
+                fun(OldReq, Req) -> queue_update(Queue, OldReq, Req) end
+            )
+        ),
+    report_metrics(Metrics, emqx_ds_beamformer:beams_n_matched(BeamMaker)),
+    %% Send out the replies:
+    ok = emqx_ds_beamformer:beams_conclude(ShardId, EndKey, BeamMaker).
+
+-spec process_batch(
+    s(),
+    emqx_ds:stream(),
+    emqx_ds:topic_filter(),
+    [{emqx_ds:message_key(), emqx_types:message()}],
+    [emqx_ds_beamformer:sub_state()],
+    emqx_ds_beamformer:beam_builder()
+) -> emqx_ds_beamformer:beam_builder().
+process_batch(_S, _Stream, _TopicFilter, [], _Candidates, Beams) ->
+    Beams;
+process_batch(S, Stream, TopicFilter, [{Key, Msg} | Rest], Candidates0, Beams0) ->
+    Candidates = queue_lookup(S, Stream, TopicFilter, Key) ++ Candidates0,
+    Beams = emqx_ds_beamformer:beams_add(Stream, Key, Msg, Candidates, Beams0),
+    process_batch(S, Stream, TopicFilter, Rest, Candidates, Beams).
+
+%% It's always worth trying to fulfill the oldest requests first,
+%% because they have a better chance of producing a batch that
+%% overlaps with other pending requests.
+%%
+%% This function implements a heuristic that tries to find such poll
+%% request. It simply compares the keys (and nothing else) within a
+%% small sample of pending polls, and picks request with the smallest
+%% key as the starting point.
```

Comment:
Q: this suggests that the oldest request has the smallest key, but since `keypos = 1` for the queue, but this field apparently is a stream according to `?queue_elem`;  wouldn't the oldest request have the smallest `StartKey` (i.e., `StartKey` should be the first element in the `?queue_elem` tuple)?